c/shared/source/uforead/uforead.c

```diff
@@ -1132,58 +1162,83 @@ static long getGlyphOrderIndex(ufoCtx h, char* glyphName) {
     newGLIFRec = dnaNEXT(h->data.glifRecs);
     newGLIFRec->glyphName = glyphName;
     newGLIFRec->glyphOrder = glyphOrder;
+    if (!h->parseState.CIDMap) {  /* do not proceed if currently parsing CIDMap in lib.plist. This information will be filled in later. */
+        if (fileName == NULL) {
+            fatal(h, ufoErrParse, "Encountered glyph reference in contents.plist with an empty file path. Text: '%s'.", getBufferContextPtr(h));
+        }
+        newGLIFRec->glifFileName = memNew(h, 1 + strlen(fileName));
+        sprintf(newGLIFRec->glifFileName, "%s", fileName);
+        newGLIFRec->altLayerGlifFileName = NULL;
+    }
+    newGLIFRec->cid = -1;
+    newGLIFRec->iFD = -1;
+    return newGLIFRec;
+}
+
+static void addGLIFRec(ufoCtx h, char* glyphName, xmlNodePtr cur) {
+    GLIF_Rec* foundGlyph = NULL;
+    char* fileName = parseXMLKeyValue(h, cur);
     if (fileName == NULL) {
         fatal(h, ufoErrParse, "Encountered glyph reference in contents.plist with an empty file path. Text: '%s'.", getBufferContextPtr(h));
     }
-    newGLIFRec->glifFileName = memNew(h, 1 + strlen(fileName));
-    sprintf(newGLIFRec->glifFileName, "%s", fileName);
-    newGLIFRec->altLayerGlifFileName = NULL;
+    int nameLength = strlen(fileName) - 5;  /* '.glif' removed */
+    char *subbuff = memNew(h, (sizeof(char) * (nameLength + 1)));
+    memcpy(subbuff, &fileName[0], nameLength);
+    subbuff[nameLength] = '\0';
+    foundGlyph = findGLIFRecByName(h, subbuff);
+    memFree(h, subbuff);
+    if (foundGlyph == NULL){
+        addNewGLIFRec(h, glyphName, cur);
+    } else {
+        foundGlyph->glifFileName = fileName;
+        if (foundGlyph->glyphOrder == -1)
+            foundGlyph->glyphOrder = getGlyphOrderIndex(h, glyphName);
+    }
 }
 
-static int findGLIFRecByName(ufoCtx h, char *glyphName)
+int glifRecNameComparator(const void *a, const void *b) {
+    const GLIF_Rec *p1 = (GLIF_Rec *) a;
+    const GLIF_Rec *p2 = (GLIF_Rec *) b;
+    return strcmp((*p1).glyphName, (*p2).glyphName);
+}
+
+static GLIF_Rec* findGLIFRecByName(ufoCtx h, char *glyphName)
 {
-    int i = 0;
-    while (i < h->data.glifRecs.cnt) {
-        GLIF_Rec* glifRec;
-        glifRec = &h->data.glifRecs.array[i];
-        if (strEqual(glifRec->glyphName, glyphName)) {
-            return i;
-        }
-        i++;
-    }
-    return -1;
+    GLIF_Rec *glifKey = memNew(h, sizeof(GLIF_Rec*));
+    glifKey->glyphName = glyphName;
+
+    GLIF_Rec* glif = bsearch(glifKey, h->data.glifRecs.array, h->data.glifRecs.cnt, sizeof(GLIF_Rec), glifRecNameComparator);
+    memFree(h, glifKey);
+    return glif;
 }
 
 static void updateGLIFRec(ufoCtx h, char* glyphName, xmlNodePtr cur) {
-    int index;
+    GLIF_Rec* glifRec;
     char* fileName = parseXMLKeyValue(h, cur);
     if (!keyValueParseable(h, cur, fileName, glyphName))
         return;
+    if (fileName == NULL) {
+        /* this is basically muted for now, as the previous check will return and skip if not parseable.
```

Comment:
I agree it sounds weird — I think my intention was to say that with a verbosity flag, the user would be told that the GLIF is missing. I'll downgrade this to a warning message instead, since in the current implementation we just ignore if this happens.